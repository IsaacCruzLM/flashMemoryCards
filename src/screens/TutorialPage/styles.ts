import {StyleSheet} from 'react-native';
import themes from '../../styles/themes';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    paddingVertical: 100,
    paddingHorizontal: 24,
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  infosContainer: {
    width: '100%',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
  },
  buttonContainer: {
    width: '100%',
  },
  title: {
    fontFamily: themes.fonts.medium.fontFamily,
    fontSize: themes.typography.fontSizeLargeTitle,
    color: themes.colors.background,
    fontWeight: '700',
    textAlign: 'center',
    marginBottom: 30,
  },
  subTitle: {
    fontFamily: themes.fonts.regular.fontFamily,
    fontSize: themes.typography.fontSizeText,
    color: themes.colors.background,
    textAlign: 'center',
    marginBottom: 70,
    fontWeight: '500',
  },
});

export default styles;
