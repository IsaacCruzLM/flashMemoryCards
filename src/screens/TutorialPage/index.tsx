import React, {useEffect, useState} from 'react';
import {View} from 'react-native';
import Animated, {
  withTiming,
  useAnimatedStyle,
  useSharedValue,
  withDelay,
} from 'react-native-reanimated';

import NavigationService from '../../navigation/NavigationService';

import LinearGradientView from '../../components/LinearGradientView';
import Button from '../../components/Button';

import InfoTip from './components/InfoTip';

import styles from './styles';

const TutorialPage = () => {
  const [isDisabled, setIsDisabled] = useState(true);
  const buttonOpacity = useSharedValue(0);

  useEffect(() => {
    setTimeout(() => setIsDisabled(false), 6500);
    buttonOpacity.value = withDelay(6000, withTiming(1, {duration: 500}));
  });

  const buttonContainerStyle = useAnimatedStyle(() => {
    return {
      opacity: buttonOpacity.value,
    };
  });

  return (
    <LinearGradientView>
      <View style={styles.container}>
        <View style={styles.infosContainer}>
          <InfoTip
            title={'Crie suas anotações'}
            subTitle={'Crie suas anotações dos mais diversos assuntos'}
            icon={'pen-plus'}
          />
          <InfoTip
            title={'Separe-as por assunto'}
            subTitle={'Melhore sua organização, separando suas anotações'}
            icon={'folder-plus'}
            animationDelay={2000}
          />
          <InfoTip
            title={'Seja relembrado'}
            subTitle={
              'Receba uma recordação para relembrar suas anotações, após 1 dia, 7 dias, 1 mês e 3 meses.'
            }
            icon={'bell-ring'}
            animationDelay={4000}
          />
        </View>
        <Animated.View style={[styles.buttonContainer, buttonContainerStyle]}>
          <Button
            disabled={isDisabled}
            onPress={() => NavigationService.navigate('Home')}
            label="Começar"
          />
        </Animated.View>
      </View>
    </LinearGradientView>
  );
};

export default TutorialPage;
