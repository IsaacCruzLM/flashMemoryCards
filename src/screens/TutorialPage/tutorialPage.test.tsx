import React, {ElementType} from 'react';
import renderer, {act, ReactTestRenderer} from 'react-test-renderer';
import AsyncStorage from '@react-native-async-storage/async-storage';
import TutorialPage from './index';

jest.mock('../../navigation/NavigationService', () => ({
  navigate: jest.fn(),
}));

jest.mock('../../databases/utils', () => {
  return {
    __esModule: true,
    default: {
      insertItemInWMDB: jest.fn(() => Promise.resolve()),
    },
  };
});

jest.mock('../../components/Button', () => {
  return {
    __esModule: true,
    default: 'Button',
  };
});

describe('TutorialPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders correctly with initial state', () => {
    let tree: ReactTestRenderer;

    act(() => {
      tree = renderer.create(<TutorialPage />);
    });

    expect(tree!.toJSON()).toMatchSnapshot();
  });

  it('enables button after timeout', () => {
    let tree: ReactTestRenderer;

    act(() => {
      tree = renderer.create(<TutorialPage />);
    });

    // Initial state - button should be disabled
    expect(
      tree!.root.findByType('Button' as ElementType).props.disabled,
    ).toBeTruthy();

    // Fast-forward time
    act(() => {
      jest.advanceTimersByTime(6500);
    });

    // Button should be enabled
    expect(
      tree!.root.findByType('Button' as ElementType).props.disabled,
    ).toBeFalsy();
  });

  it('handles button press correctly', async () => {
    const NavigationService = require('../../navigation/NavigationService');
    const WmdbUtils = require('../../databases/utils').default;

    let tree: ReactTestRenderer;

    await act(async () => {
      tree = renderer.create(<TutorialPage />);
    });

    // Enable button
    await act(async () => {
      jest.advanceTimersByTime(6500);
    });

    // Press button and wait for all promises to resolve
    await act(async () => {
      const button = tree!.root.findByType('Button' as ElementType);
      await button.props.onPress();
      await Promise.resolve();
      jest.runAllTimers();
    });

    // Wait for a tick to ensure all promises are settled
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
    });

    // Verify AsyncStorage was called
    expect(AsyncStorage.setItem).toHaveBeenCalledWith('tutorialSeen', 'true');

    // Verify database insertions
    expect(WmdbUtils.insertItemInWMDB).toHaveBeenCalledTimes(2);
    expect(WmdbUtils.insertItemInWMDB).toHaveBeenCalledWith('categories', {
      color: '#000000',
      createdAt: expect.any(Date),
      icon: 'archive-edit',
      name: 'Minhas Anotações',
    });
    expect(WmdbUtils.insertItemInWMDB).toHaveBeenCalledWith('subjects', {
      color: '#00c65c',
      createdAt: expect.any(Date),
      name: 'Geral',
    });

    // Verify navigation
    expect(NavigationService.navigate).toHaveBeenCalledWith('Home');
  });

  it('initializes animations correctly', () => {
    const reanimated = require('react-native-reanimated');

    act(() => {
      renderer.create(<TutorialPage />);
    });

    expect(reanimated.useSharedValue).toHaveBeenCalledWith(0);

    // Verify each call with its specific delay
    expect(reanimated.withDelay).toHaveBeenCalledTimes(4);
    expect(reanimated.withDelay).toHaveBeenNthCalledWith(1, 0, undefined);
    expect(reanimated.withDelay).toHaveBeenNthCalledWith(2, 2000, undefined);
    expect(reanimated.withDelay).toHaveBeenNthCalledWith(3, 4000, undefined);
    expect(reanimated.withTiming).toHaveBeenCalledWith(1, {duration: 500});
  });
});
