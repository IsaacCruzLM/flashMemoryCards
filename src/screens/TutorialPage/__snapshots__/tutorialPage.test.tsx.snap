// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TutorialPage renders correctly with initial state 1`] = `
<View
  colors={
    Array [
      "rgba(34, 39, 46, 0.7)",
      "rgba(34, 39, 46, 1)",
    ]
  }
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-between",
        "paddingHorizontal": 24,
        "paddingVertical": 100,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "marginBottom": 40,
            "width": "100%",
          }
        }
      >
        <View
          style={[Function]}
        >
          <Icon
            color="#FFFFFF"
            name="pen-plus"
            size={70}
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "700",
                  "marginBottom": 8,
                  "textAlign": "center",
                },
                [Function],
              ]
            }
          >
            Crie suas anotações
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "textAlign": "center",
                },
                [Function],
              ]
            }
          >
            Crie suas anotações dos mais diversos assuntos
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "marginBottom": 40,
            "width": "100%",
          }
        }
      >
        <View
          style={[Function]}
        >
          <Icon
            color="#FFFFFF"
            name="folder-plus"
            size={70}
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "700",
                  "marginBottom": 8,
                  "textAlign": "center",
                },
                [Function],
              ]
            }
          >
            Separe-as por assunto
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "textAlign": "center",
                },
                [Function],
              ]
            }
          >
            Melhore sua organização, separando suas anotações
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "marginBottom": 40,
            "width": "100%",
          }
        }
      >
        <View
          style={[Function]}
        >
          <Icon
            color="#FFFFFF"
            name="bell-ring"
            size={70}
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "700",
                  "marginBottom": 8,
                  "textAlign": "center",
                },
                [Function],
              ]
            }
          >
            Seja relembrado
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "textAlign": "center",
                },
                [Function],
              ]
            }
          >
            Receba uma recordação para relembrar suas anotações, após 1 dia, 7 dias, 1 mês e 3 meses.
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "width": "100%",
          },
          [Function],
        ]
      }
    >
      <Button
        disabled={true}
        label="Começar"
        onPress={[Function]}
      />
    </View>
  </View>
</View>
`;
